### DISCUSSION TITLE
Why osgEarth render map with elevation layer so slowly?

### DISCUSSION THREAD

**Discussion by smarttigerfly** [pelicanmapping/osgearth#2274] [Support Q&A] [answered]
When I render map with elevation layer using osgearth, it run very slowly, and the result is this:
![2023-04-21-22-26-34](https://user-images.githubusercontent.com/13800186/233662834-903d9226-8e0a-4a90-98cf-83e79f712389.png)
The `main.cpp` is:
```
#include <osgViewer/Viewer>
#include <osgEarth/MapNode>
#include <osgEarth/TMS>
#include <osgEarth/EarthManipulator>
#include <osgEarth/Composite>
#include <osgEarth/GeodeticGraticule>
#include <osgEarth/GLUtils>

using namespace osgEarth;
using namespace osgEarth::Util;

int main(int argc, char** argv)
{
    osg::setNotifyLevel(osg::WARN);
    osgEarth::setNotifyLevel(osg::NotifySeverity::INFO);

    osgEarth::initialize();
    osg::ArgumentParser arguments(&argc, argv);
    // initialize a viewer:
    osgViewer::Viewer viewer(arguments);
    EarthManipulator* manip = new EarthManipulator();
    viewer.setCameraManipulator(manip);

    // create the empty map.
    Map* map = new Map();
    map->setProfile(Profile::create(Profile::GLOBAL_GEODETIC));

    //ImageLayer
    TMSImageLayer* imageLayer = new TMSImageLayer();
    std::string filename("E:\\osgEarthData\\global_TMS\\tms.xml");  
    imageLayer->setURL(filename.c_str());

    //ElevationLayer
    TMSElevationLayer* elev1 = new TMSElevationLayer();
    elev1->setURL("E:\\osgEarthData\\china_TMS\\Elevation\\beijing\\tms.xml");
    elev1->setTileSize(32);
    map->addLayer(elev1);


    map->addLayer(imageLayer);
    map->addLayer(elev1);

    GeodeticGraticule* gr = new GeodeticGraticule();
    map->addLayer(gr);

    // make the map scene graph:
    MapNode* node = new MapNode(map);
    node->open();  
    node->getTerrainOptions().setMinLOD(1);
    node->getTerrainOptions().setHeightFieldSkirtRatio(0.08);
    node->getTerrainOptions().setGPUTessellation(true);
    node->getTerrainOptions().setPriorityScale(3.0);
    node->getTerrainOptions().setConcurrency(8);
    node->getTerrainOptions().setMinNormalMapLOD(10);
    node->getTerrainOptions().setProgressive(true);

    viewer.setSceneData(node);

    // disable the small-feature culling
    viewer.getCamera()->setSmallFeatureCullingPixelSize(-1.0f);
    // set a near/far ratio that is smaller than the default. This allows us to get
   // closer to the ground without near clipping. If you need more, use --logdepth
    viewer.getCamera()->setNearFarRatio(0.0001);
    GLUtils::setGlobalDefaults(viewer.getCamera()->getOrCreateStateSet());
    viewer.setRealizeOperation(new GL3RealizeOperation());

    viewer.realize();
    // zoom to a good startup position
    manip->setViewpoint(Viewpoint(
        "Home",
        116.52, 40.35, 10.88,   // longitude, latitude, altitude
        0, -9, 73100.49), // heading, pitch, range
        5.0);                    // duration

   return viewer.run();
  
}
```

**Comment by smarttigerfly** [comment]
When I build the example application in `release` mode, and it run as expected.
But the display content has background color between different tiles, how to eliminate background color gaps between different tiles?
![2023-04-24-15-42-30-2](https://user-images.githubusercontent.com/13800186/233952315-0c2e5c64-41f8-4c9c-b475-2905e1528cef.png)
The gaps may be due to, the near screen display the high LOD tile, the far screen display the low LOD tile, can force the LOD to be the same?

**Reply by remoe** [reply]
Try to use ```skirt_ratio```. It does not exists in current documentation of osgEarth. Here is an old documentation about it:

https://github.com/gwaldron/osgearth/blob/7d9604ebfc56a16225de5d92b57c4fbc173e2f49/docs/old_rst/source/references/drivers/terrain/mp.rst

**Reply by smarttigerfly** [reply]
I want to have a try, but can't built the `mp` terrain driver in `osgearth 3.3`.  The `rex` driver options has the attribute `heightFieldSkirtRatio` which can be set by `node->getTerrainOptions().setHeightFieldSkirtRatio(0.08);`. It's function seems to be the same with `skirt_ratio`, because there is `conf.set( "skirt_ratio", heightFieldSkirtRatio() );` in `TerrainOptions.cpp`. But the `node->getTerrainOptions().setHeightFieldSkirtRatio(0.08);` has no effect on result. Set the 'RangeMode' to be `osg::LOD::RangeMode::PIXEL_SIZE_ON_SCREEN` through `node->getTerrainOptions().setRangeMode(osg::LOD::RangeMode::PIXEL_SIZE_ON_SCREEN);`, the result will be better, but still have many the background color gaps between different tiles

**Comment by jasonbeverage** [comment, marked as answer]
How did you build your elevation dataset?  You have to take great care to make sure the edges of neighboring tiles share elevation values or you will get the gaps like you are seeing.  osgEarth's GDALElevation layer will take care of that for you or look at our commercial ReadyMap server.  Using an image based tool like gdal2tiles won't work correctly.

**Reply by smarttigerfly** [reply]
The elevation dataset is served by local TMS elevation tiles, which may be generated by `gdal2tiles`. So you may be right. I will have a try to use other elevation dataset.

**Reply by smarttigerfly** [reply]
Yes, you are right. Thank you very much!
When I use the `.tif` elevation dataset, the application runs correctly!
![2023-04-24-22-53-36](https://user-images.githubusercontent.com/13800186/234035616-70e02358-6b2f-4d64-ad3d-6b8d4209cc1c.png)

And, another question, the `GeodeticGraticule`'s lines can or not to be set `AltitudeSymbol::CLAMP_ABSOLUTE`, I didn't find the style set function in class `GeodeticGraticule`.
